### PR TITLE
OS-243 Audit api to fetch audit log from DB

### PR DESCRIPTION
**Pull request by sknirmalkar89**

### CODE REVIEW

File: java/registry/src/main/java/io/opensaber/registry/controller/RegistryController.java [hunk 1]
```diff
@@ -130,15 +136,31 @@
...
-    public ResponseEntity<Response> fetchAudit(@PathVariable("id") String id) {
-        ResponseParams responseParams = new ResponseParams();
-        Response response = new Response(Response.API_ID.AUDIT, "OK", responseParams);
-        // if (auditEnabled) {
+	@RequestMapping(value = "/audit", method = RequestMethod.POST)
+	public ResponseEntity<Response> fetchAudit() {
+		ResponseParams responseParams = new ResponseParams();
+		Response response = new Response(Response.API_ID.AUDIT, "OK", responseParams);
+		JsonNode payload = apiMessage.getRequest().getRequestMapNode();
+		if (auditEnabled && Constants.DATABASE.equals(auditStoreType)) {
```
Comment (indrajra) [inline java/registry/src/main/java/io/opensaber/registry/controller/RegistryController.java]: Else, set the response that auditing store is not a database.
You can be creative and say 405 - method not allowed as well.

File: java/registry/src/main/java/io/opensaber/registry/helper/RegistryHelper.java [hunk 2]
```diff
@@ -170,5 +178,90 @@ public String updateEntity(JsonNode inputJson, String userId) throws Exception {
...
+		JsonNode resultNode = searchService.search(auditNode);
+		return resultNode;
+
+	}
+
+	private JsonNode getSearchQueryNodeForAudit(JsonNode inputJson) throws Exception {
+
+		ObjectMapper mapper = new ObjectMapper();
+
+		JsonNode entityTypeNodeArr = inputJson.get("entityType");
```
Comment (indrajra) [inline java/registry/src/main/java/io/opensaber/registry/helper/RegistryHelper.java]: Create an audit Pojo and mark mandatory or null checks there. Individual if conditions make the function bigger.

File: java/registry/src/main/java/io/opensaber/registry/helper/RegistryHelper.java [hunk 3]
```diff
@@ -170,5 +178,90 @@ public String updateEntity(JsonNode inputJson, String userId) throws Exception {
...
+		ArrayNode newEntityArrNode = mapper.createArrayNode();
+		for (JsonNode entity : entityTypeNodeArr) {
+			newEntityArrNode.add(entity.asText() + "_Audit");
+		}
+		ObjectNode searchNode = mapper.createObjectNode();
+
+		searchNode.set("entityType", newEntityArrNode);
+		ObjectNode filterNode = mapper.createObjectNode();
+
+		if (null != inputJson.get("action") && !inputJson.get("action").asText().isEmpty()) {
```
Comment (indrajra) [inline java/registry/src/main/java/io/opensaber/registry/helper/RegistryHelper.java]: You're basically using an audit payload and constructing a filter query (jsonNode). Please put this outside somewhere like an AuditHelper, not in the main class.
RegistryHelper is the top level - class that must act like a set of commands and so must be simple to read.

File: java/registry/src/main/resources/audit_frame.json [hunk 4]
```diff
@@ -1,8 +1,77 @@
 {
-  "@context": {
-    "rdf": "http://www.w3.org/1999/02/22-rdf-syntax-ns#",
-    "rdfs": "http://www.w3.org/2000/01/rdf-schema#",
-    "teacher": "http://localhost:8080/",
-    "xsd": "http://www.w3.org/2001/XMLSchema#"
+  "$schema": "http://json-schema.org/draft-07/schema",
```
Comment (indrajra) [inline java/registry/src/main/resources/audit_frame.json]: I suggest we avoid using the word frame. It is used heavily in Json-LD and suggest we call raw_audit_schema.json.

File: java/registry/src/main/resources/audit_frame.json [hunk 5]
```diff
@@ -1,8 +1,77 @@
...
+  "_osConfig": {
+    "osComment": ["This section contains the OpenSABER specific configuration information",
+      "privateFields: Optional; list of field names to be encrypted and stored in database",
+      "signedFields: Optional; list of field names that must be pre-signed",
+      "indexFields: Optional; list of field names used for creating index. Enclose within braces to indicate it is a composite index. In this definition, (serialNum, teacherCode) is a composite index and teacherName is a single column index.",
+      "uniqueIndexFields: Optional; list of field names used for creating unique index. Field names must be different from index field name",
+      "systemFields: Optional; list of fields names used for system standard information like created, updated timestamps and userid"],
+
+
+      "systemFields": []
```
Comment (indrajra) [inline java/registry/src/main/resources/audit_frame.json]: I think this must be on - audit also must carry created, updated timestamps.

File: java/registry/src/main/java/io/opensaber/registry/helper/RegistryHelper.java [hunk 6]
```diff
@@ -170,5 +178,90 @@ public String updateEntity(JsonNode inputJson, String userId) throws Exception {
...
+
+		JsonNode entityTypeNodeArr = inputJson.get("entityType");
+		if(null == entityTypeNodeArr) {
+			throw new NullPointerException("entityType cannot be null");
+		}
+		if (!entityTypeNodeArr.isArray()) {
+			throw new CustomException("entityType should be an array");
+		}
+		if (entityTypeNodeArr.size() <= 0) {
+			throw new CustomException("entityType should not be an empty array");
```
Comment (indrajra) [inline java/registry/src/main/java/io/opensaber/registry/helper/RegistryHelper.java]: Can you create an Audit exception - that has further sub exception classes like, audit id missing, invalid arguments.

File: java/registry/src/main/java/io/opensaber/registry/service/ElasticSearchService.java [hunk 7]
```diff
@@ -78,7 +78,7 @@ public JsonNode search(JsonNode inputQueryNode) throws IOException {
                 resultNode.set(indexName, node);
                 if(node !=  null) {
                     AuditInfo auditInfo = new AuditInfo();
-                    auditInfo.setOp(Constants.AUDIT_ACTION_SEARCH_OP);
+                    auditInfo.setOp(actionSearch.toLowerCase());
```
Comment (indrajra) [inline java/registry/src/main/java/io/opensaber/registry/service/ElasticSearchService.java]: why toLowerCase() ?

File: java/registry/src/main/java/io/opensaber/registry/util/AuditHelper.java [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+import io.opensaber.registry.exception.audit.InvalidArguementException;
+import io.opensaber.registry.middleware.util.Constants;
+import io.opensaber.pojos.FilterOperators;
+
+@Component
+public class AuditHelper {
+
+	private static Logger logger = LoggerFactory.getLogger(AuditHelper.class);
+
+	public JsonNode getSearchQueryNodeForAudit(JsonNode inputJson, String uuidPropertyName)
```
Comment (indrajra) [inline java/registry/src/main/java/io/opensaber/registry/util/AuditHelper.java]: can you break this further? - all the validations separately, lines 55-97 to construct the filter node. This improves readability and fixing easier.

Comment (sknirmalkar89) [reply java/registry/src/main/java/io/opensaber/registry/util/AuditHelper.java]: sure

File: java/registry/src/main/java/io/opensaber/registry/service/ElasticSearchService.java (hunk 7 above)
Comment (sknirmalkar89) [reply java/registry/src/main/java/io/opensaber/registry/service/ElasticSearchService.java]: earlier it was lowercase i.e the reason it is lowercase

File: java/registry/src/main/java/io/opensaber/registry/util/AuditHelper.java (hunk 8 above)
Comment (sknirmalkar89) [reply java/registry/src/main/java/io/opensaber/registry/util/AuditHelper.java]: Done

**Pull request merged**